lib/shopify_cli/tunnel.rb

```diff
@@ -52,205 +39,85 @@ def stop(ctx)
-    rescue
-      []
-    end
+    def url(ctx)
+      return nil unless running?
 
-    ##
-    # Returns Boolean if a tunnel is running on a given port
-    #
-    # #### Parameters
-    #
-    # * `port` - port to check
-    #
-    # #### Returns
-    #
-    # * true / false
-    #
-    def running_on?(port)
-      extract_port = ->(tunnel) { URI(tunnel.dig(*TUNNEL_ADDRESS_KEY_PATH)).port }
-      matches_port = ->(occupied_port) { occupied_port == port }
-      stats.fetch(TUNNELS_FIELD, []).map(&extract_port).any?(&matches_port)
-    rescue
-      false
+      process = ShopifyCLI::ProcessSupervision.for_ident(:tunnel)
+      fetch_url(ctx, process.log_path)
     end
 
     private
 
-    def install(ctx)
-      ngrok = "ngrok#{ctx.executable_file_extension}"
-      return if File.exist?(ngrok_path(ctx))
-      check_prereq_command(ctx, "curl")
-      check_prereq_command(ctx, ctx.linux? ? "unzip" : "tar")
-      spinner = CLI::UI::SpinGroup.new
-      spinner.add(ctx.message("core.tunnel.installing")) do
-        zip_dest = File.join(ShopifyCLI.cache_dir, "ngrok.zip")
-        unless File.exist?(zip_dest)
-          ctx.system("curl", "-o", zip_dest, DOWNLOAD_URLS[ctx.os], chdir: ShopifyCLI.cache_dir)
-        end
-        args = if ctx.linux?
-          %W(unzip -u #{zip_dest})
-        else
-          %W(tar -xf #{zip_dest})
-        end
-        ctx.system(*args, chdir: ShopifyCLI.cache_dir)
-        ctx.rm(zip_dest)
-      end
-      spinner.wait
-
-      # final check to see if ngrok is accessible
-      unless File.exist?(ngrok_path(ctx))
-        ctx.abort(ctx.message("core.tunnel.error.ngrok", ngrok, ShopifyCLI.cache_dir))
-      end
+    def running?
+      ShopifyCLI::ProcessSupervision.running?(:tunnel)
     end
 
     def fetch_url(ctx, log_path)
-      LogParser.new(log_path)
+      LogParser.new(log_path).url
     rescue RuntimeError => e
       stop(ctx)
       raise e.class, e.message
     end
 
-    def ngrok_path(ctx)
-      ngrok = "ngrok#{ctx.executable_file_extension}"
-      File.join(ShopifyCLI.cache_dir, ngrok)
-    end
-
-    def seconds_to_hm(seconds)
-      format("%d hours %d minutes", seconds / 3600, seconds / 60 % 60)
-    end
-
-    def start_ngrok(ctx, port)
-      ngrok_command = "\"#{ngrok_path(ctx)}\" http -inspect=false -log=stdout -log-level=debug #{port}"
-      process = ShopifyCLI::ProcessSupervision.start(:ngrok, ngrok_command)
-      log = fetch_url(ctx, process.log_path)
-      [log.url, log.account]
-    end
-
-    def restart_ngrok(ctx, port)
-      ShopifyCLI::ProcessSupervision.stop(:ngrok)
-      start_ngrok(ctx, port)
-    end
-
-    def check_prereq_command(ctx, command)
-      cmd_path = ctx.which(command)
-      ctx.abort(ctx.message("core.tunnel.error.prereq_command_required", command)) if cmd_path.nil?
-      ctx.done(ctx.message("core.tunnel.prereq_command_location", command, cmd_path))
+    def start_tunnel(ctx, port)
+      # check_prereq_command(ctx, "npx")
```

Comment:
The check for the npm dependency is still pending. Would this be a good place? Or maybe we should check Node before, as we do now when you try to create a Node app for example? Are there people using the CLI without running `serve`?

Reply:
> The check for the npm dependency is still pending. Would this be a good place? Or maybe we should check Node before, as we do now when you try to create a Node app for example? Are there people using the CLI without running serve?

I'd do it at the command level, same as controllers in Rails to prevent invalid states from entering the system. Once passed the command, the code can safely assume that Node is available in the system.